Update ERC-4337: Add 'preVerificationGas' estimation explainer (#955)
ethereum/ERCs: ERCS/erc-4337.md

To prevent DoS attacks on bundlers, they must make sure the validation methods above pass the validation rules, which constrain their usage of opcodes and storage.
For the complete procedure see [ERC-7562](./eip-7562.md)

### Estimating `preVerificationGas`

This document does not specify a canonical way to estimate this value,
as it depends on non-permanent network properties such as operation and data gas pricing and the expected bundle size.

However, the requirement is for the estimated value to be sufficient to cover the following costs:

* Base bundle transaction cost. On Ethereum, `21000` gas divided by the number of `UserOperations`.
* The calldata gas cost related to the `UserOperation`.
* Static `EntryPoint` contract code execution.

> * Cost of performing the memory expansion when loading the `UserOperation` into EVM memory.
>   This includes the call to the `innerHandleOp()` function which is a major part of the `EntryPoint` implementation.

Reviewer: ```suggestion
* Static memory cost when loading the fixed size fields of the `UserOperation` into EVM memory
* Memory cost (including expansion cost) due to contexts returned by paymaster `postOp` function, if relevant.
  * External call to the `innerHandleOp()` function which is a major part of the `EntryPoint` implementation.
```

  Note that this value is not static and depends on the `UserOperation`'s position in the bundle.
* EIP-7702 authorization cost, if any.

> 

Reviewer: ```suggestion
The bundler SHOULD require a slack in PreVerificationGas value, to accommodate changes in bundle size (and expected memory expansion cost in the bundle), and the expected position of the `UserOperation` in the bundle.
```

To prevent DoS attacks on bundlers, they must make sure the validation methods above pass the validation rules, which constrain their usage of opcodes and storage.
For the complete procedure see [ERC-7562](./eip-7562.md)

### Estimating `preVerificationGas`

This document does not specify a canonical way to estimate this value,
as it depends on non-permanent network properties such as operation and data gas pricing and the expected bundle size.

However, the requirement is for the estimated value to be sufficient to cover the following costs:

* Base bundle transaction cost. On Ethereum, `21000` gas divided by the number of `UserOperations`.
* The calldata gas cost related to the `UserOperation`.
* Static `EntryPoint` contract code execution.
* Static memory cost when loading the fixed size fields of the `UserOperation` into EVM memory
* Memory cost (including expansion cost) due to context returned by paymaster `validatePaymasterUserOp` function, if relevant.
  * External call to the `innerHandleOp()` function which is a major part of the `EntryPoint` implementation.

>   Note that this value is not static and depends on the `UserOperation`'s position in the bundle.

Reviewer 2: ```suggestion
  Note that this value is not static and depends on the number of `UserOperation`s in the bundle, and the total paymaster context sizes they return
```

* EIP-7702 authorization cost, if any.

> The bundler MUST require a slack in `PreVerificationGas` value, to accommodate memory expansion costs, affected by total size of contexts of previous `UserOperations` in the future bundle,  and the expected position of the `UserOperation` in the bundle.

Reviewer: ```suggestion
The bundler MUST require a slack in `PreVerificationGas` value, to accommodate memory expansion costs in the future bundle,  and the expected position of the `UserOperation` in it.
```
